src/lib/util/opresolv.h

```diff
@@ -102,61 +136,98 @@ class option_resolution
+		entry(const option_guide::entry &guide_entry);
+
+		// returns true if this entry is subject to range constraints
+		bool is_ranged() const { return !m_ranges.empty(); }
+
+		// accessors
+		bool is_pertinent() const { return m_is_pertinent; }
+		const std::string &value() const;
+		int value_int() const;
+		const std::string &default_value() const { return m_default_value; }
+		const rangelist &ranges() const { return m_ranges; }
+		const option_guide::entrylist::const_iterator enum_value_begin() const { return m_enum_value_begin; }
+		const option_guide::entrylist::const_iterator enum_value_end() const { return m_enum_value_end; }
+
+		// accessors for guide data
+		option_guide::entry::option_type option_type() const { return m_guide_entry.type(); }
+		const char *display_name() const { return m_guide_entry.display_name(); }
+		const char *identifier() const { return m_guide_entry.identifier(); }
+		int parameter() const { return m_guide_entry.parameter(); }
+
+
+		// mutators
+		bool set_value(const std::string &value);
+		bool set_value(int value) { return set_value(numeric_value(value)); }
+
+		// higher level operations
+		bool can_bump_lower() const;
+		bool can_bump_higher() const;
+		bool bump_lower();
+		bool bump_higher();
+
+	private:
+		// references to the option guide
+		const option_guide::entry &				m_guide_entry;
+		option_guide::entrylist::const_iterator	m_enum_value_begin;
+		option_guide::entrylist::const_iterator	m_enum_value_end;
+
+		// runtime state
+		bool									m_is_pertinent;
+		std::string								m_value;
+		std::string								m_default_value;
+		rangelist								m_ranges;
+
+		// methods
+		void parse_specification(const char *specification);
+		void set_enum_value_range(option_guide::entrylist::const_iterator begin, option_guide::entrylist::const_iterator end);
+		static std::string numeric_value(int value);
+		rangelist::const_iterator find_in_ranges(int value) const;
+	};
 
-	// accessors
-	const char *specification() const { return m_specification; }
-	const option_guide *find_option(int option_char) const;
-	const option_guide *index_option(int indx) const;
+	// ctor/dtor
+	option_resolution(const option_guide &guide);
+	~option_resolution();
 
-	// processing option guides
-	static size_t count_options(const option_guide *guide, const char *specification);
+	// sets a specification
+	void set_specification(const std::string &specification);
 
-	// processing option specifications
-	static error list_ranges(const char *specification, int option_char,
-		range *range, size_t range_count);
-	static error get_default(const char *specification, int option_char, int *val);
-	static error is_valid_value(const char *specification, int option_char, int val);
-	static bool contains(const char *specification, int option_char);
+	// entry lookup - note that we're not exposing m_entries directly because we don't really
+	// have a way of making the list be immutable but the members be mutable at the same time
```

Comment:
So make a custom proxy iterator class or something - don't do it in a half-arsed way.